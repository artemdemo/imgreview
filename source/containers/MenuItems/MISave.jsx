import React from 'react';
import { connect } from 'react-redux';
import Icon from '../../components/Icon/Icon';
import Popup from '../../components/Popup/Popup';
import MainMenuItem from '../../components/MainMenu/MainMenuItem';
import { saveCanvas } from '../../model/canvas/canvasActions';
import { blurShapes } from '../../model/shapes/shapesActions';

class MISave extends React.PureComponent {
    constructor(props) {
        super(props);

        this.popupRef = React.createRef();
    }

    onClick = () => {
        const { blurShapes } = this.props;
        blurShapes();
        this.popupRef.current.show();
    };

    render() {
        const { canvas } = this.props;
        return (
            <React.Fragment>
                <MainMenuItem
                    onClick={this.onClick}
                    disabled={canvas.image == null}
                >
                    <Icon
                        name='floppy-o'
                    />
                </MainMenuItem>
                <Popup
                    ref={this.popupRef}
                    buttons={[
                        {
                            className: 'btn btn-secondary',
                            text: 'Cancel',
                        },
                        {
                            className: 'btn btn-primary',
                            text: 'Save',
                            onClick: () => {
                                const { saveCanvas } = this.props;
                                const { value } = this.fileNameRef.current;
                                if (value !== '') {
                                    saveCanvas(value);
                                }
                            },
                        },
                    ]}
                    showCloseBtn={false}
                >
                    <div className='form-group'>
                        <label htmlFor='saveAs'>Save as (*.png)</label>
                        <input
                            className='form-control'
                            placeholder='Enter file name'
                            ref={this.fileNameRef}
                            id='saveAs'
                        />
                    </div>
                </Popup>
            </React.Fragment>
        );
    }
}

export default connect(
    state => ({
        canvas: state.canvas,
    }), {
        blurShapes,
        saveCanvas,
    },
)(MISave);
