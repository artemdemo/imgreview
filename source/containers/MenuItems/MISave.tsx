/* eslint-disable jsx-a11y/label-has-associated-control,jsx-a11y/label-has-for,react/no-unused-state */
import React from 'react';
import { connect } from 'react-redux';
import { Form, Field } from 'react-final-form';
import { TReduxState } from '../../reducers';
import Icon from '../../components/Icon/Icon';
import Popup from '../../components/Popup/Popup';
import FormGroup from '../../components/FormGroup/FormGroup';
import FormInput from '../../components/FormInput/FormInput';
import MainMenuItem from '../../components/MainMenu/MainMenuItem';
import Button from '../../components/Button/Button';
import FormButtonsRow from '../../components/FormButtonsRow/FormButtonsRow';
import PopupButtonsContainer from '../../components/Popup/PopupButtonsContainer';
import { saveCanvas, TSaveCanvas } from '../../model/canvas/canvasActions';
import { TStateCanvas } from '../../model/canvas/canvasReducer';

type Props = {
    canvas: TStateCanvas;
    showColorPicker: () => void;
    saveCanvas: TSaveCanvas
};

class MISave extends React.PureComponent<Props> {
    private readonly popupRef: any;
    private readonly nameRef: any;

    constructor(props) {
        super(props);

        this.state = {
            name: '',
        };
        this.popupRef = React.createRef();
        this.nameRef = React.createRef();
    }

    onCancel = () => {
        this.popupRef.current.hide();
    };

    onClick = () => {
        const { canvas } = this.props;
        this.setState({
            name: canvas.imageOriginName,
        }, () => {
            this.popupRef.current.show();
        });
    };

    onPopupOpen = () => {
        this.nameRef.current.focus();
    };

    onSubmit = (values) => {
        const { saveCanvas } = this.props;
        const { name } = values;
        if (name !== '') {
            saveCanvas(name.trim());
            this.setState({
                name: '',
            });
            this.popupRef.current.hide();
        }
    };

    render() {
        const { canvas } = this.props;
        return (
            <React.Fragment>
                <MainMenuItem
                    onClick={this.onClick}
                    disabled={canvas.image == null}
                >
                    <Icon
                        name='floppy-o'
                        title='Save'
                    />
                </MainMenuItem>
                <Popup
                    ref={this.popupRef}
                    onSubmit={this.onSubmit}
                    showCloseBtn={false}
                    onOpen={this.onPopupOpen}
                >
                    <Form
                        initialValues={this.state}
                        onSubmit={this.onSubmit}
                        validate={(values: any) => {
                            const errors: any = {};
                            if (!values.name || values.name.replace(/\s/g, '') === '') {
                                errors.name = 'Name can\'t be empty';
                            }
                            return errors;
                        }}
                        render={({ handleSubmit, invalid }) => (
                            <form onSubmit={handleSubmit}>
                                <Field
                                    name='name'
                                    render={({ input, meta }) => (
                                        <FormGroup
                                            errorText={meta.error}
                                        >
                                            <label htmlFor='saveAs'>Save as (*.png)</label>
                                            <FormInput
                                                placeholder='Enter file name'
                                                id='saveAs'
                                                ref={this.nameRef}
                                                {...input}
                                            />
                                        </FormGroup>
                                    )}
                                />
                                <PopupButtonsContainer>
                                    <FormButtonsRow>
                                        <Button
                                            onClick={this.onCancel}
                                            secondary
                                        >
                                            Cancel
                                        </Button>
                                        <Button
                                            type='submit'
                                            disabled={invalid}
                                        >
                                            Save
                                        </Button>
                                    </FormButtonsRow>
                                </PopupButtonsContainer>
                            </form>
                        )}
                    />
                </Popup>
            </React.Fragment>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        canvas: state.canvas,
    }), {
        saveCanvas,
    },
)(MISave);
