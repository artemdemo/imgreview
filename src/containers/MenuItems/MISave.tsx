/* eslint-disable jsx-a11y/label-has-associated-control,jsx-a11y/label-has-for,react/no-unused-state */
import React from 'react';
import { connect } from 'react-redux';
import { Form, Field } from 'react-final-form';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faDownload } from '@fortawesome/free-solid-svg-icons';
import { TReduxState } from '../../reducers';
import Popup from '../../components/Popup/Popup';
import FormGroup from '../../components/FormGroup/FormGroup';
import FormInput from '../../components/FormInput/FormInput';
import TopMenuItem from '../../components/TopMenu/TopMenuItem';
import Button, { EButtonAppearance } from '../../components/Button/Button';
import FormButtonsRow from '../../components/FormButtonsRow/FormButtonsRow';
import PopupButtonsContainer from '../../components/Popup/PopupButtonsContainer';
import { TStateCanvas } from '../../model/canvas/canvasReducer';
import * as canvasApi from '../../../srcCanvas/api';
import * as gaService from '../../services/ganalytics';
import { t } from '../../services/i18n';

type TProps = {
  canvas: TStateCanvas;
  showColorPicker: () => void;
  disabled: boolean;
};

type TState = {
  name: string;
};

class MISave extends React.PureComponent<TProps, TState> {
  static readonly defaultProps = {
    disabled: false,
  };

  state = {
    name: '',
  };

  onCancel = () => {
    // this.popupRef.current?.hide();
  };

  onClick = () => {
    const { canvas } = this.props;
    this.setState(
      {
        name: canvas.imageOriginName,
      },
      () => {
        // this.popupRef.current?.show();
      }
    );

    gaService.sendEvent({
      eventCategory: gaService.EEventCategories.MenuClick,
      eventAction: gaService.EEventActions.SaveImage,
    });
  };

  onPopupOpen = () => {
    // this.nameRef.current?.focus();
  };

  onPopupClose = () => {
    this.setState({
      name: '',
    });
  };

  onSubmit = (values: any) => {
    const { name } = values;
    if (name !== '') {
      canvasApi.exportCanvasToImage(name.trim());
      // this.popupRef.current?.hide();
    }
  };

  onValidate = (values: { name?: string }) => {
    const errors: { name?: string } = {};
    if (!values.name || values.name.replace(/\s/g, '') === '') {
      errors.name = "Name can't be empty";
    }
    return errors;
  };

  renderForm = ({ handleSubmit, invalid }: any) => (
    <form onSubmit={handleSubmit}>
      <Field
        name="name"
        render={({ input, meta }) => (
          <FormGroup errorText={meta.error}>
            <label htmlFor="saveAs">Save as (*.png)</label>
            <FormInput placeholder="Enter file name" id="saveAs" {...input} />
          </FormGroup>
        )}
      />
      <PopupButtonsContainer>
        <FormButtonsRow>
          <Button
            onClick={this.onCancel}
            appearance={EButtonAppearance.SECONDARY}
          >
            Cancel
          </Button>
          <Button type="submit" disabled={invalid}>
            Save
          </Button>
        </FormButtonsRow>
      </PopupButtonsContainer>
    </form>
  );

  render() {
    const { disabled } = this.props;
    return (
      <>
        <TopMenuItem
          onClick={this.onClick}
          title={t('menu.save')}
          disabled={disabled}
          stopPropagation={false}
        >
          <FontAwesomeIcon icon={faDownload} />
        </TopMenuItem>
        <Popup
          showCloseBtn={false}
          onOpen={this.onPopupOpen}
          onClose={this.onPopupClose}
        >
          <Form
            initialValues={this.state}
            onSubmit={this.onSubmit}
            validate={this.onValidate}
            render={this.renderForm}
          />
        </Popup>
      </>
    );
  }
}

export default connect(
  (state: TReduxState) => ({
    canvas: state.canvas,
  }),
  {}
)(MISave);
