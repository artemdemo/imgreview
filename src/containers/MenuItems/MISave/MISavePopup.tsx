import React, { useEffect, useState } from 'react';
import FormGroup from '../../../components/FormGroup/FormGroup';
import FormInput from '../../../components/FormInput/FormInput';
import PopupButtonsContainer from '../../../components/Popup/PopupButtonsContainer';
import FormButtonsRow from '../../../components/FormButtonsRow/FormButtonsRow';
import Button, { EButtonAppearance } from '../../../components/Button/Button';
import Popup from '../../../components/Popup/Popup';

type Props = {
  onSubmit: (name: string) => void;
  onCancel: () => void;
  show: boolean;
  nameInit: string;
};

export const MISavePopup: React.FC<Props> = (props) => {
  const { onCancel, onSubmit, nameInit, show } = props;
  const [name, setName] = useState('');
  const [nameError, setNameError] = useState('');

  useEffect(() => {
    if (show) {
      setName(nameInit);
    }
  }, [show]);

  useEffect(() => {
    if (name.replace(/\s/g, '') === '') {
      setNameError("Name can't be empty");
    } else {
      setNameError('');
    }
  }, [name]);

  const onPopupClose = () => {
    setName('');
    setNameError('');
    onCancel();
  };

  return (
    <Popup showCloseBtn={false} onClose={onPopupClose} show={show}>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          onSubmit(name);
        }}
      >
        <FormGroup errorText={nameError}>
          <label htmlFor="saveAs">Save as (*.png)</label>
          <FormInput
            placeholder="Enter file name"
            id="saveAs"
            value={name}
            onChange={(e) => {
              setName(e.target.value);
            }}
          />
        </FormGroup>
        <PopupButtonsContainer>
          <FormButtonsRow>
            <Button onClick={onCancel} appearance={EButtonAppearance.SECONDARY}>
              Cancel
            </Button>
            <Button type="submit" disabled={nameError !== ''}>
              Save
            </Button>
          </FormButtonsRow>
        </PopupButtonsContainer>
      </form>
    </Popup>
  );
};
