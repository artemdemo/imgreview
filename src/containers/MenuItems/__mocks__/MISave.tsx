import React from 'react';

const MISave = props => (
    <div data-mock='MISave'>
        {props.children}
    </div>
);

export default MISave;
