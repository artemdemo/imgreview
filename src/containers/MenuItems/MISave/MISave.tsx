/* eslint-disable jsx-a11y/label-has-associated-control,jsx-a11y/label-has-for,react/no-unused-state */
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { TReduxState } from '../../../reducers';
import { TopMenuItem } from '../../../components/TopMenu/TopMenuItem';
import * as canvasApi from '../../../../srcCanvas/api';
import * as gaService from '../../../services/ganalytics';
import { t } from '../../../services/i18n';
import { MISavePopup } from './MISavePopup';
import { TStateCanvas } from '../../../model/canvas/canvasReducer';
import { EIcon, ImgIcon } from '../ImgIcon/ImgIcon';

type Props = {
  disabled: boolean;
};

export const MISave: React.FC<Props> = (props) => {
  const { disabled } = props;
  const [nameInit, setNameInit] = useState('');
  const [showPopup, setShowPopup] = useState(false);
  const canvas = useSelector<TReduxState, TStateCanvas>(
    (state) => state.canvas
  );

  const onClick = () => {
    setNameInit(canvas.imageOriginName);
    setShowPopup(true);
    gaService.sendEvent({
      eventCategory: gaService.EEventCategories.MenuClick,
      eventAction: gaService.EEventActions.SaveImage,
    });
  };

  const onPopupClose = () => {
    setShowPopup(false);
  };

  const onSubmit = (name: string) => {
    if (name !== '') {
      canvasApi.exportCanvasToImage(name.trim());
      setShowPopup(false);
    }
  };

  return (
    <>
      <TopMenuItem
        onClick={onClick}
        title={t('menu.save')}
        disabled={disabled}
        stopPropagation={false}
      >
        <ImgIcon icon={EIcon.save} />
      </TopMenuItem>
      <MISavePopup
        onSubmit={onSubmit}
        onCancel={onPopupClose}
        show={showPopup}
        nameInit={nameInit}
      />
    </>
  );
};
