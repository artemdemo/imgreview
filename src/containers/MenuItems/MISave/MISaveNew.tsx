/* eslint-disable jsx-a11y/label-has-associated-control,jsx-a11y/label-has-for,react/no-unused-state */
import React, { useState } from 'react';
import { TopMenuItem } from '../../../components/TopMenu/TopMenuItem';
import * as canvasApi from '../../../../srcCanvas/api';
import * as gaService from '../../../services/ganalytics';
import { t } from '../../../services/i18n';
import { MISavePopup } from './MISavePopup';
import { EIcon, ImgIcon } from '../ImgIcon/ImgIcon';
import { exportCanvasToImageNew } from '../../../../srcCanvas/api';

type Props = {
  disabled?: boolean;
};

export const MISaveNew: React.FC<Props> = (props) => {
  const { disabled = false } = props;
  const [showPopup, setShowPopup] = useState(false);

  const onClick = () => {
    // setShowPopup(true);
    // gaService.sendEvent({
    //   eventCategory: gaService.EEventCategories.MenuClick,
    //   eventAction: gaService.EEventActions.SaveImage,
    // });
    canvasApi.exportCanvasToImageNew('aaa');
  };

  const onPopupClose = () => {
    setShowPopup(false);
  };

  const onSubmit = (name: string) => {
    if (name !== '') {
      canvasApi.exportCanvasToImage(name.trim());
      setShowPopup(false);
    }
  };

  return (
    <>
      <TopMenuItem
        onClick={onClick}
        title={t('menu.save')}
        disabled={disabled}
        stopPropagation={false}
      >
        SaveNew
      </TopMenuItem>
      <MISavePopup
        onSubmit={onSubmit}
        onCancel={onPopupClose}
        show={showPopup}
        nameInit={''}
      />
    </>
  );
};
