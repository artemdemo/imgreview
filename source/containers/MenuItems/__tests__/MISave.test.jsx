import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MISave from '../MISave';

jest.mock('react-redux');
jest.mock('../../../canvas/Arrow/Arrow');
jest.mock('../../../components/Popup/Popup');
jest.mock('../../../components/Icon/Icon');
jest.mock('../../../components/MainMenu/MainMenuItem');

describe('MISave', () => {
    jest.clearAllMocks();
    const reactReduxMock = require('react-redux');

    it('should render', () => {
        const state = {
            canvas: {
                image: true,
            },
        };
        const tree = renderer.create(
            <MISave
                canvas={state.canvas}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
        expect(reactReduxMock.__getLastMaps().mapStateToProps(state)).toEqual({
            canvas: state.canvas,
        });
    });

    it('should be disabled', () => {
        const state = {
            canvas: {
                image: null,
            },
        };
        const tree = renderer.create(
            <MISave
                canvas={state.canvas}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should trigger popup open and blurShapes', () => {
        const state = {
            canvas: {
                image: true,
            },
        };
        const blurShapesMock = jest.fn();
        const wrapper = mount(
            <MISave
                canvas={state.canvas}
                blurShapes={blurShapesMock}
            />
        );
        const instance = wrapper.instance();
        const showMock = jest.fn();
        instance.popupRef = {
            current: {
                show: showMock,
            },
        };
        instance.onClick();
        expect(showMock).toBeCalled();
        expect(blurShapesMock).toBeCalled();
    });

    it('should handle popup open', () => {
        const imageOriginName = 'some name';
        const state = {
            canvas: {
                imageOriginName,
                image: true,
            },
        };
        const wrapper = mount(
            <MISave
                canvas={state.canvas}
            />
        );
        const instance = wrapper.instance();
        const setStateSpy = jest.spyOn(instance, 'setState');
        instance.onPopupOpen();
        expect(setStateSpy).toBeCalledWith(
            expect.objectContaining({
                name: imageOriginName,
            }),
            expect.any(Function),
        );
    });
});
