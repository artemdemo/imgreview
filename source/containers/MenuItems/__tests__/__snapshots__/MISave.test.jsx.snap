// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MISave should be disabled 1`] = `
Array [
  <div
    data-mock="TopMenuItem"
    onClick={[Function]}
  >
    <div
      data-mock="Icon"
    />
  </div>,
  <div
    data-mock="Popup"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="sc-bdVaJa gBePXc"
      >
        <label
          htmlFor="saveAs"
        >
          Save as (*.png)
        </label>
        <input
          className="form-input"
          id="saveAs"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter file name"
          value=""
        />
        <div
          className="sc-bwzfXH hwXLHn"
        >
          Name can't be empty
        </div>
      </div>
      <div
        className="popup-buttons-container"
      >
        <div
          className="sc-ifAKCX rijn"
        >
          <button
            className="sc-htpNat sc-bxivhb fQkliK"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
        <div
          className="sc-ifAKCX rijn"
        >
          <button
            className="sc-htpNat sc-bxivhb fEUrCp"
            disabled={true}
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </form>
  </div>,
]
`;

exports[`MISave should render 1`] = `
Array [
  <div
    data-mock="TopMenuItem"
    onClick={[Function]}
  >
    <div
      data-mock="Icon"
    />
  </div>,
  <div
    data-mock="Popup"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="sc-bdVaJa gBePXc"
      >
        <label
          htmlFor="saveAs"
        >
          Save as (*.png)
        </label>
        <input
          className="form-input"
          id="saveAs"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter file name"
          value=""
        />
        <div
          className="sc-bwzfXH hwXLHn"
        >
          Name can't be empty
        </div>
      </div>
      <div
        className="popup-buttons-container"
      >
        <div
          className="sc-ifAKCX rijn"
        >
          <button
            className="sc-htpNat sc-bxivhb fQkliK"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
        <div
          className="sc-ifAKCX rijn"
        >
          <button
            className="sc-htpNat sc-bxivhb fEUrCp"
            disabled={true}
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </form>
  </div>,
]
`;
